src/libraries/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/HttpConnectionPool.cs

```diff
@@ -1640,55 +1751,150 @@ public void ReturnConnection(HttpConnection connection)
+                    // We do this after the above in order to satisfy any requests that were queued before the pool was disposed of.
+                    Debug.Assert(_associatedHttp2ConnectionCount > (_availableHttp2Connections?.Count ?? 0));
+                    _associatedHttp2ConnectionCount--;
+                    poolDisposed = true;
+                }
+                else if (usable)
+                {
+                    if (_availableHttp2Connections is null)
                     {
-                        Array.Copy(localHttp2Connections, invalidatedIndex + 1, newHttp2Connections, invalidatedIndex, newHttp2Connections.Length - invalidatedIndex);
+                        _availableHttp2Connections = new List<Http2Connection>();
                     }
 
-                    _http2Connections = newHttp2Connections;
+                    // Add connection to the pool.
+                    _availableHttp2Connections.Add(connection);
+                    if (NetEventSource.Log.IsEnabled()) connection.Trace("Put HTTP/2 connection in pool.");
+                    return;
                 }
             }
+
+            if (poolDisposed)
+            {
+                if (NetEventSource.Log.IsEnabled()) connection.Trace("Disposing HTTP/2 connection returned to pool. Pool was disposed.");
+                connection.Dispose();
+                return;
+            }
+
+            Debug.Assert(!usable);
+
+            // We need to wait until the connection is usable again.
+            DisableHttp2Connection(connection);
+        }
+
+        /// <summary>
+        /// Disable usage of the specified connection because it cannot handle any more streams at the moment.
+        /// We will register to be notified when it can handle more streams (or becomes permanently unusable).
+        /// </summary>
+        private void DisableHttp2Connection(Http2Connection connection)
+        {
+            if (NetEventSource.Log.IsEnabled()) connection.Trace("");
+
+            Task.Run(async () =>
+            {
+                bool usable = await connection.WaitForAvailableStreamsAsync().ConfigureAwait(false);
+
+                if (NetEventSource.Log.IsEnabled()) connection.Trace($"WaitForAvailableStreamsAsync completed, {nameof(usable)}={usable}");
+
+                if (usable)
+                {
+                    ReturnHttp2Connection(connection, isNewConnection: false);
+                }
+                else
+                {
+                    // Connection has shut down.
+                    lock (SyncObj)
+                    {
+                        Debug.Assert(_availableHttp2Connections is null || !_availableHttp2Connections.Contains(connection));
+                        Debug.Assert(_associatedHttp2ConnectionCount > 0);
+
+                        _associatedHttp2ConnectionCount--;
+
+                        CheckForHttp2ConnectionInjection();
```

Comment:
Another cyclic call chain that look suspicious to me `CheckForHttp2ConnectionInjection` -> `AddHttp2ConnectionAsync` -> `ReturnHttp2Connection` -> `DisableHttp2Connection` -> `CheckForHttp2ConnectionInjection`.